core/integration/secret_test.go

```diff
@@ -186,3 +188,113 @@ func (SecretSuite) TestBigScrubbed(ctx context.Context, t *testctx.T) {
+			WithNewFile("main.go", `package main
+
+import (
+	"context"
+	"dagger/foo/internal/dagger"
+	"fmt"
+	"time"
+)
+
+type Foo struct{}
+
+
+// This function sets the value of secret in vault, gets its plaintext value. Then
+// this function updates the value of secret in vault (to simulate expired or changed secret), and sleeps for 20s to allow for ttl (if any) 
+// to expire. It then gets its plaintext value again.
+// After that it returns both the values as string, which our tescase then verifies.
+func (m *Foo) VerifySecret(ctx context.Context, vault *dagger.Service, secret *dagger.Secret) (string, error) {
+	_, err := dag.Container().From("hashicorp/vault").
+		WithEnvVariable("VAULT_ADDR", "http://vault:8200").
+		WithEnvVariable("VAULT_TOKEN", "vault-root-token").
+		WithServiceBinding("vault", vault).
+		WithExec([]string{"sh", "-c", "vault kv put secret/my-secret username=\"admin\" password=\"original-password\""}).Sync(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	original, err := secret.Plaintext(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	// simulate an update in secret while the pipeline is running
+	_, err = dag.Container().From("hashicorp/vault").
+		WithEnvVariable("VAULT_ADDR", "http://vault:8200").
+		WithEnvVariable("VAULT_TOKEN", "vault-root-token").
+		WithServiceBinding("vault", vault).
+		WithExec([]string{"sh", "-c", "vault kv put secret/my-secret username=\"admin\" password=\"updated-password\""}).Sync(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	// wait for ttl to expire, to simulate a long running process
+	time.Sleep(20 * time.Second)
+
+	// now the pipeline needs secret again.
+	updated, err := secret.Plaintext(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("original: %s\nupdated: %s", original, updated), nil
+}
+`).
+			With(daggerCall("-vvv", "verify-secret", fmt.Sprintf("--secret=%s", secretURL), "--vault=tcp://vault:8200")).Stdout(ctx)
+	}
+
+	testcases := []struct {
+		name                    string
+		secret                  string
+		expectedUpdatedPassword string
+	}{
+		{
+			name:                    "without any ttl it caches forever",
+			secret:                  "vault://data/my-secret.password",
+			expectedUpdatedPassword: "original-password",
+		},
+		{
+			name:                    "respect ttl if provided",
+			secret:                  "vault://data/my-secret.password?ttl=10s",
+			expectedUpdatedPassword: "updated-password",
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(ctx context.Context, t *testctx.T) {
+			c := connect(ctx, t)
+
+			vault := c.Container().
+				From("hashicorp/vault").
+				WithEnvVariable("VAULT_DEV_ROOT_TOKEN_ID", "vault-root-token").
+				WithEnvVariable("VAULT_LOG_LEVEL", "debug").
+				WithExposedPort(8200).
+				AsService(dagger.ContainerAsServiceOpts{
+					UseEntrypoint:                 true,
+					ExperimentalPrivilegedNesting: true,
+					InsecureRootCapabilities:      true,
+				})
```

Comment:
why do we create a vault service here  if we're not accessing it outside the module? wouldn't it be better if we just keep this withing the module test?

Reply:
to test the secret fetching functionality, we need to simulate a vault service running on host system.